gas.md

```diff
@@ -0,0 +1,554 @@
+    ```hs
+    computationCost(LOCALOP rREG) = localStateCost + addressSize*copyWordCost
+    estimatedResultSize(LOCALOP rREG) = addressSize
+    ```
+* `TMESTAMP rREG`
+    ```hs
+    computationCost(LOCALOP rREG) = localStateCost + timeStampSize*copyWordCost
+    estimatedResultSize(LOCALOP rREG) = timeStampSize
+    ```
+* `DIFFICULTY rREG`
+    ```hs
+    computationCost(LOCALOP rREG) = localStateCost + difficultySize()*copyWordCost
+    estimatedResultSize(LOCALOP rREG) = difficultySize()
+    ```
+* `CALLVALUE rREG`
+    ```hs
+    computationCost(CALLVALUE rREG) = localStateCost + callValueSize()*copyWordCost
+    estimatedResultSize(CALLVALUE rREG) = callValueSize()
+    ```
+* `BLOCKHASH rREG wN`
+  - Computation cost is a constant; result size is *blockHashSize*
+    ```hs
+    computationCost(CALLVALUE rREG) = blockHashCost + blockHashSize*copyWordCost
+    estimatedResultSize(CALLVALUE rREG) = blockHashSize
+    ```
+
+#### JUMPs
+
+* `JUMPDEST`
+    ```hs
+    computationCost(JUMPDEST(_)) = jumpDestCost
+    ```
+* `JUMP`
+    ```hs
+    computationCost(JUMP(_)) = jumpCost
+    ```
+* `JUMPI`
+    ```hs
+    computationCost(JUMPI(_) cREG)
+      | not (isZero cREG) = bitCost + jumpCost
+      | otherwise         = skipJumpiCost
+    ```
+* `LOCALCALL`
+  Save the local context (including the current register stack memory
+  requirements), copy the arguments into the new context and jump to
+  the call site.
+  ```hs
+  computationCost(LOCALCALL(_, nARGS, _) _ rARGS) =
+    saveContextCost + wordCopyCost * (sum [registerSize r | r <- rARGS]) + jumpCost
+  ```
+
+* `RETURN`
+  Copy values from return registers to caller locations restore local context,
+  including (the current register stack memory requirements), mark registers'
+  data for reclaiming, and jump back to call site. Note that, since the
+  callee registers' data will be reclaimed, copying the return data in the
+  callers' registers is not needed; just metadata needs to be copied.
+  ```hs
+  computationCost(RETURN(nRETURNS) rVALUES) =
+    restoreContextCost + metadataCopyCost * nRETURNS + jumpCost
+  requiredRegisterMemory = callerRegisterMemory +
+    sum [registerLoadDelta(r, v) | (r, v) <- getReturns() `zip` rVALUES]
+  ```
+
+#### `STOP` and `REVERT`
+
+* `STOP`
+  ```hs
+  computationCost(STOP) = stopCost() -- defined in terms of the current execution context
+  ```
+* `REVERT`
+  ```hs
+  computationCost(REVERT(nRETURNS) rVALUES) = wordCopyCost * registerSize(rValues) +
+    revertCost() -- defined in terms of the current execution context.
+  ```
+
+TODO:  Figure out more precise costs for these
+
+
+#### Logging
+We use the same schema from the yellow paper, but taking into account the size
+of the logged registers.
+
+* `LOG`
+  ```hs
+  computationCost(LOG0 rMEMSTART rMEMWIDTH) =
+    logCost + value rMEMWIDTH * logDataCost
+  computationCost(LOG1 rMEMSTART rMEMWIDTH rW0) =
+    logCost + value rMEMWIDTH * logDataCost + registerSize rW0 * logTopicWordCost
```

Comment:
do we actually want unbounded log topics? I'm unsure.

Reply:
Note that we bound memory indices by 2^256 so there is precedent here

Reply:
I'm not sure we want unbounded log topics either. My first guess is that they have potential to create some bad system issues, so we don't want them. However,  from a gas computation perspective, this should not be relevant. If rWi are limited in size in the IELE semantics, I could use the same formula or I could put a flat fee here, as EVM does.